webserver-plugins/plugin-hibernate/src/main/java/org/webpieces/plugin/hibernate/PersistenceHelper.java

```diff
@@ -0,0 +1,133 @@
+package org.webpieces.plugin.hibernate;
+
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.EntityTransaction;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tags;
+
+import org.webpieces.plugin.hibernate.metrics.DatabaseMetric;
+import org.webpieces.plugin.hibernate.metrics.DatabaseTransactionTags;
+import org.webpieces.util.context.Context;
+import org.webpieces.util.context.ContextKey;
+
+@Singleton
+public class PersistenceHelper {
+
+    private final EntityManagerFactory factory;
+    private final TxCompleters txCompleters;
+    private final MeterRegistry meterRegistry;
+
+    @Inject
+    public PersistenceHelper(EntityManagerFactory factory, TxCompleters txCompleters, MeterRegistry meterRegistry) {
+        this.factory = factory;
+        this.txCompleters = txCompleters;
+        this.meterRegistry = meterRegistry;
+    }
+
+    public void execute(String executionId, Consumer<EntityManager> consumer) {
+        execute(executionId, entityManager -> {
+            consumer.accept(entityManager);
+            return null;
+        });
+    }
+
+    public <T> T execute(String executionId, Function<EntityManager, T> function) {
+
+        if (Em.get() != null) {
+            throw new IllegalStateException("Cannot open another entityManager when are already have one open");
+        }
+
+        return executeImpl(executionId, function);
+
+    }
+
+    private <T> T executeImpl(String executionId, Function<EntityManager, T> function) {
+
+        EntityManager mgr = factory.createEntityManager();
+        Em.set(mgr);
+
+        long start = System.currentTimeMillis();
+
+        try {
+            T result = function.apply(mgr);
+            mgr.close();
+            return result;
+        } catch (Throwable t) {
+            txCompleters.closeEm(t, mgr);
+            throw t;
+        } finally {
+            Em.set(null);
+            long end = System.currentTimeMillis();
+            monitorExecutionTime(executionId, start, end);
+        }
+
+    }
+
+    public void executeTransaction(String executionId, Consumer<EntityManager> consumer) {
+        executeTransaction(executionId, entityManager -> {
+            consumer.accept(entityManager);
+            return null;
+        });
+    }
+
+    public <T> T executeTransaction(String executionId, Function<EntityManager, T> function) {
+
+        if(Em.get() == null) {
+            return execute(executionId, (Function<EntityManager,T>)(em) -> executeTransactionImpl(executionId, function));
```

Comment:
`(Function<EntityManager,T>)` This cast should not be necessary

Reply:
It shouldn't be but IntelliJ gets upset so it's there.